src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITypeParameterObjectCreationTests.cs

```diff
@@ -0,0 +1,318 @@
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void TypeParameterObjectCreation_WithCollectionInitializer()
+        {
+            string source = @"
+using System.Collections.Generic;
+
+class C1
+{
+    void M<T>(T t1) where T : C1, IEnumerable<int>, new()
+    {
+        t1 = /*<bind>*/new T() { 1 }/*</bind>*/;
+    }
+    void Add(int i) { }
+}
+";
+            var expectedDiagnostics = DiagnosticDescription.None;
+
+            string expectedOperationTree = @"
+ITypeParameterObjectCreationOperation (OperationKind.TypeParameterObjectCreation, Type: T) (Syntax: 'new T() { 1 }')
+  Initializer: 
+    IObjectOrCollectionInitializerOperation (OperationKind.ObjectOrCollectionInitializer, Type: T) (Syntax: '{ 1 }')
+      Initializers(1):
+          IInvocationOperation ( void C1.Add(System.Int32 i)) (OperationKind.Invocation, Type: System.Void, IsImplicit) (Syntax: '1')
+            Instance Receiver: 
+              IInstanceReferenceOperation (ReferenceKind: ImplicitReceiver) (OperationKind.InstanceReference, Type: T, IsImplicit) (Syntax: 'T')
+            Arguments(1):
+                IArgumentOperation (ArgumentKind.Explicit, Matching Parameter: i) (OperationKind.Argument, Type: null, IsImplicit) (Syntax: '1')
+                  ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+                  InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+                  OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
+";
+            VerifyOperationTreeAndDiagnosticsForTest<ObjectCreationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void TypeParameterObjectCreation_NoNewConstraint()
+        {
+            string source = @"
+using System.Collections;
+using System.Collections.Generic;
+
+class C1
+{
+    void M<T>(T t1, bool b) where T : C1, IEnumerable<int>
+    {
+        t1 = /*<bind>*/new T() { 1, b ? 2 : 3 }/*</bind>*/;
+    }
+    void Add(int i) { }
+}
+";
+            var expectedDiagnostics = new DiagnosticDescription[] {
+                // CS0304: Cannot create an instance of the variable type 'T' because it does not have the new() constraint
+                //         t1 = /*<bind>*/new T() { 1, b ? 2 : 3 }/*</bind>*/;
+                Diagnostic(ErrorCode.ERR_NoNewTyvar, "new T() { 1, b ? 2 : 3 }").WithArguments("T").WithLocation(9, 24)
+            };
+
+            // Asserts that no ITypeParameterObjectCreationOperation is generated in this tree
+            string expectedOperationTree = @"
+IInvalidOperation (OperationKind.Invalid, Type: T, IsInvalid) (Syntax: 'new T() { 1, b ? 2 : 3 }')
```

Comment:
Consider adding a CFG test for this scenario

Reply:
I didn't consider it relevant since it's not a type parameter object creation, this test was simply to confirm that behavior. I can add one to the invalidoperation tests though.

---
In reply to: [193162689](https://github.com/dotnet/roslyn/pull/27460#discussion_r193162689) [](ancestors = 193162689)